engine/table/src/main/java/io/deephaven/engine/table/impl/UngroupOperation.java

```diff
@@ -431,214 +332,420 @@ private class UngroupListener extends BaseTable.ListenerImpl {
-            // transform things for the key columns)
-            final RowSet modifiedRowSet = ungroupModifiedModified.build();
-            final TableUpdateImpl downstream = new TableUpdateImpl(addedRowSet, removedRowSet, modifiedRowSet,
-                    shiftBuilder.build(), ModifiedColumnSet.ALL);
+            resultRowset.insert(added);
+
+            setNewBase(newBase);
+
+            final RowSet modifiedRowSet = modifiedByModifiesBuilder.build();
+            transformer.clearAndTransform(upstream.modifiedColumnSet(), result.getModifiedColumnSetForUpdates());
+            final TableUpdateImpl downstream = new TableUpdateImpl(added, removedRowSet, modifiedRowSet,
+                    shiftBuilder.build(), result.getModifiedColumnSetForUpdates());
             result.notifyListeners(downstream);
         }
 
-        private void rebase(final int newBase) {
-            final RowSetBuilderSequential builder = RowSetFactory.builderSequential();
-            final long[] sizes = new long[parentRowset.intSize()];
-            computeMaxSize(parentRowset, sizes, false);
-            getUngroupIndex(sizes, builder, newBase, parentRowset);
-            final WritableRowSet newRowSet = builder.build();
-            rebase(newBase, newRowSet);
+        private class ShiftWrapper {
+            RowSetShiftData shiftData;
+            int pos = -1;
+            long nextKey;
+
+            ShiftWrapper(final RowSetShiftData shiftData) {
+                this.shiftData = shiftData;
+                if (shiftData.nonempty()) {
+                    pos = 0;
+                    nextKey = shiftData.getBeginRange(pos);
+                } else {
+                    nextKey = RowSet.NULL_ROW_KEY;
+                }
+            }
+
+            long currentShiftDelta() {
+                return shiftData.getShiftDelta(pos);
+            }
+
+            /**
+             * Advance the shift data to the given key.
+             * 
+             * @param advanceTo the key to advance to
+             * @return true if the shift data is positioned at the desired key, false otherwise
+             */
+            boolean advance(final long advanceTo) {
+                if (nextKey == RowSet.NULL_ROW_KEY) {
+                    // already at the end
+                    return false;
+                }
+                if (advanceTo <= nextKey) {
+                    // already at the right place
+                    return advanceTo == nextKey;
+                }
+
+                if (advanceTo <= shiftData.getEndRange(pos)) {
+                    Assert.geq(advanceTo, "advanceTo", shiftData.getBeginRange(pos), "shiftData.getBeginRange(pos)");
+                    nextKey = advanceTo;
+                    return true;
+                }
+
+                while (pos < shiftData.size()) {
+                    if (advanceTo <= shiftData.getEndRange(pos)) {
+                        nextKey = Math.max(shiftData.getBeginRange(pos), nextKey);
```

Comment:
We think this is a bug:
```suggestion
                        nextKey = Math.max(shiftData.getBeginRange(pos), advanceTo);
```